### ISSUE TITLE
Move pending users/orgs to separate LDAP branches

### ISSUE THREAD

**Issue by landryb** [georchestra/georchestra#2108]
As discussed on https://groups.google.com/forum/#!topic/georchestra/dWD9BC7xGzM

It would be simpler to have pending users and orgs in separate LDAP branches, as external services might auth users on the `ou=users` or synchronize data from the `ou=orgs` branch, and they would have to be modified to specialcase the PENDING ones..

Right now, from testing & looking at the LDAP, to me the current rules are : 
  * a pending user is defined by an entry in `ou=users`, this user is member of the group `cn=PENDING` under `ou=roles` and of the org `cn=my_org` under `ou=orgs`.
  *  validating an user removes him from the 'PENDING' role and adds him to the 'USER' role
  * a pending org is defined by two entries in `ou=orgs`, an `organization` object and a `groupOfMembers` object, the latter having the 'PENDING' value for `businessCategory`
  * validating an org changes the 'PENDING' value for 'REGISTERED'

I would propose the following changes:
  * a pending user is defined by an entry in `ou=pendingusers`,
    * if he created a new org, this user is a member of an org `cn=my_org` under `ou=pendingorgs`
    * if he requested to be linked to an existing org, a temporary attribute in the user LDAP entry stores the DN of this org (or the org display page tweaks from #2080 needs to distinguish the pending users from the regular users)
  * validating this user moves (MODRDN) the corresponding LDAP object to the `ou=users` branch, and adds him to the 'USER' role
  * if he requested to be linked to an existing org, he is added as a member of the given org (for which the value was stored in the temporary attribute)

  * a pending org is defined by two entries in `ou=pendingorgs`
  * validating an org moves (MODRDN) the two entries to `ou=orgs`
    * membership info for the pending users are updated accordingly (should be automagically done in LDAP)
    * one needs to be able to rename the org before validating it (cf #2107, this is a case we always use..)

There are probably other constraints to enforce but right now i cant see more.. and the migration path involves:
  * migration notes:
    * create the new ou entries
    * ensure no user/org is in the PENDING group/state
    * remove the PENDING role
  * updating the `georchestra.ldif` sample (while here, add a pending user/group as an example it..)
  * other things ?

In terms of UI plumbing, this requires adaptation to the pending users & orgs pages that would fetch their info from the corresponding branch, and the 'complete' list of users/orgs would have to fetch the info from both branches, distinguishing them like now (pending users in italic).

**Comment by landryb** [comment]
Playing a bit with the idea & starting with the groups->roles/orgs migration, the console really needs to be adapted to only list users from the *given* `ou=users` branch - what if you have several branches with `inetOrgPerson` ? This is my case here for various reasons (a branch holding distinct accounts for a distinct service), and in the console i only expect to see the `ou=users` branch (and `ou=pendingusers` but clearly separated) and not the other branches which georchestra has no business with..

**Comment by landryb** [comment]
it seems this sould already be the case, ie there's `userSearchBaseDN=ou=users` in the config (and it *seems* used in the AccountDaoImpl) , but strangely i see users being listed from other branches.. maybe because the filter doesnt use `userSearchBaseDN` but only `objectClass=person` in https://github.com/georchestra/georchestra/blob/master/console/src/main/java/org/georchestra/console/ds/AccountDaoImpl.java#L345 ? Dunno if this method is used to list all users, but that might explain why i see users from other branches.

**Comment by cmangeat** [comment]
Do we have to postpone cohenrency / validation checks at users or orgs  moves form pending to effective ?

Or can we consider that users/orgs in the pending branchs does book email addresses or org name ?

At this time, when an user request for an new account, there are high level checks which are performed at validation time, taking current ldap base state into account, e.g. an email address which is already used cannot be set a second time (for another user).

**Comment by fvanderbiest** [comment]
Before I forget, CAS is currently aware of the pending users, see #581
This won't be required once this enhancement is done.

**Comment by fvanderbiest** [comment]
> Do we have to postpone cohenrency / validation checks at users or orgs moves form pending to effective ?

IMO, when a user registers, we should prevent him from registering with an existing login or an existing email (at least from the validated users, but checking against all emails, both pending and valid users would be better).
Are there others checks currently performed?

**Comment by landryb** [comment]
> I'm wondering if we really want to stick to LDAP as persistence backend for pending users / orgs...
> The console already has a db schema for such things (tokens...).

I think it's better/simpler to stick to ldap, validating a user/org is just moving ldap objects between branches via MODRDN, and you can reuse all the existing validation framework made for LDAP, instead of recreating it to target the database for pending users/org.

**Comment by fvanderbiest** [comment]
sounds legit... just have to find a temporary attribute in the user LDAP entry to store the DN of this org...

**Comment by landryb** [comment]
`person` LDAP objectClass has `seeAlso` :)

**Comment by cmangeat** [comment]
if no moderation, when requesting an user creation with a new org, don't we have to set new org status to registred ?

considering

https://github.com/georchestra/georchestra/blob/master/console/src/main/java/org/georchestra/console/ws/newaccount/NewAccountFormController.java#L274

https://github.com/georchestra/georchestra/blob/master/console/src/main/java/org/georchestra/console/ws/newaccount/NewAccountFormController.java#L235

**Comment by cmangeat** [comment]
Existing validation framework has to be adapted, e.g.,

https://github.com/georchestra/georchestra/blob/master/console/src/main/java/org/georchestra/console/ds/AccountDaoImpl.java#L157-L165

https://github.com/georchestra/georchestra/blob/master/console/src/main/java/org/georchestra/console/ds/AccountDaoImpl.java#L346

In this case findByEmail has to target both 'registred' and pending user branches (not just restricting to userSearchBaseDN).

**Comment by landryb** [comment]
I dunno how the java spring ldap libs used here work, but afaik you can't easily target two distinct branches in a single request, maybe only by specifying a very complex ldap filter...

**Comment by landryb** [comment]
Did the following tests now that #2184 was merged:
1) create pending user, linked to an existing org => object created under `ou=pendingusers`
  * the user appears right away in the regular member list for the existing org, dunno if that is to be expected (it feels confusing)
  * validated said user => object migrated to `ou=users` as expected

2) create another pending user, delete it => object created under `ou=pendingusers`, then removed from it

3) create another pending user, modify its attrs in the admin console => attrs correctly updated on the object under `ou=pendingusers` - the admin page switched the top red section from ' Utilisateur en attente [confirmer]' to 'Utilisateur en attente — Confirmer d’abord l’organisme : ', i think that's a buglet.

4) create another pending user, this time trying to submit a new org with shortname BLAH => `o=BLAH` & `cn=BLAH` are correctly created under `ou=pendingorgs`; but then the console code went in an insane infinite loop creating new connections, hammering the ldap server which queries like:
```
Jan 21 16:16:54 demo slapd[15919]: conn=4678 op=1 SRCH base="cn=BLAH1,ou=orgs,dc=georchestra,dc=org" scope=0 deref=3 filter="(objectClass=*)"
...
Jan 21 16:16:54 demo slapd[15919]: conn=4679 op=1 SRCH base="cn=BLAH1,ou=pendingorgs,dc=georchestra,dc=org" scope=0 deref=3 filter="(objectClass=*)"
...
Jan 21 16:16:54 demo slapd[15919]: conn=4680 op=1 SRCH base="cn=BLAH2,ou=orgs,dc=georchestra,dc=org" scope=0 deref=3 filter="(objectClass=*)"
...
Jan 21 16:16:57 demo slapd[15919]: conn=6676 op=1 SRCH base="cn=BLAH1000,ou=orgs,dc=georchestra,dc=org" scope=0 deref=3 filter="(objectClass=*)"
....
Jan 21 16:16:57 demo slapd[15919]: conn=6677 op=1 SRCH base="cn=BLAH1000,ou=pendingorgs,dc=georchestra,dc=org" scope=0 deref=3 filter="(objectClass=*)"
...
Jan 21 16:17:52 demo slapd[15919]: conn=29451 op=1 SRCH base="cn=BLAH12387,ou=pendingorgs,dc=georchestra,dc=org" scope=0 deref=3 filter="(objectClass=*)"
```
12k connections in 1mn, nice, but a bit crazy. I think some java code went wrong somewhere :)

note that when submitting an account, it is created in the LDAP but sending the email notification triggers an NPE every time, shown to the user - i dunno if that's a fallout from #2184 but that feels suspicious:
```
==> /srv/log/console.log <==
2019-01-21 16:06:23 GlobalExceptionHandler [ERROR] java.lang.NullPointerException
        at java.util.regex.Matcher.appendReplacement(Matcher.java:804)
        at java.util.regex.Matcher.replaceAll(Matcher.java:955)
        at java.lang.String.replaceAll(String.java:2223)
        at org.georchestra.console.mailservice.Email.send(Email.java:142)
        at org.georchestra.console.mailservice.Email.send(Email.java:135)
        at org.georchestra.console.mailservice.EmailFactory.sendNewAccountRequiresModerationEmail(EmailFactory.java:138)
        at org.georchestra.console.ws.newaccount.NewAccountFormController.create(NewAccountFormController.java:288)
```

**Comment by landryb** [comment]
modifying an attr on a pending org:
 * change postalAddress correctly updates the o object under ou=pendingorgs
```
Jan 21 16:31:18 demo slapd[15919]: conn=29477 op=1 MOD dn="o=BLAH,ou=pendingorgs,dc=georchestra,dc=org"
Jan 21 16:31:18 demo slapd[15919]: conn=29477 op=1 MOD attr=postalAddress
```
  * changing short name from BLAH to BLAHFOO (ie renaming the two ldap org objects) also does the right thing in the right ldap branch:
```
Jan 21 16:36:22 demo slapd[15919]: conn=29483 op=1 MODRDN dn="cn=BLAH,ou=pendingorgs,dc=georchestra,dc=org"
Jan 21 16:36:22 demo slapd[15919]: conn=29485 op=1 MOD dn="cn=BLAHFOO,ou=pendingorgs,dc=georchestra,dc=org"
Jan 21 16:36:22 demo slapd[15919]: conn=29485 op=1 MOD attr=ou seeAlso
Jan 21 16:36:22 demo slapd[15919]: conn=29486 op=1 MODRDN dn="o=BLAH,ou=pendingorgs,dc=georchestra,dc=org"

**Comment by landryb** [comment]
Validating the pending org also does the right thing, both objects are moved to the `ou=orgs` branch:
```
Jan 21 16:40:54 demo slapd[15919]: conn=29501 op=1 MODRDN dn="cn=BLAHFOO,ou=pendingorgs,dc=georchestra,dc=org"
Jan 21 16:40:54 demo slapd[15919]: conn=29504 op=1 MODRDN dn="o=BLAHFOO,ou=pendingorgs,dc=georchestra,dc=org"
```
and seeAlso is correctly updated in the move:
```
ou=orgs,~ > cat cn=BLAHFOO

dn: cn=BLAHFOO,ou=orgs,dc=georchestra,dc=org
objectClass: top
objectClass: groupOfMembers
cn: BLAHFOO
description: 42289,42181,42153
o: testo
ou: BLAHFOO
seeAlso: o=BLAHFOO,ou=orgs,dc=georchestra,dc=org
```

**Comment by cmangeat** [comment]
Thank you for testing.
Commit you targetted is now https://github.com/georchestra/georchestra/pull/2408/commits/76d3e57797a9a80340ac029a51be27abafc75a4d.

**Comment by landryb** [comment]
Retested now with the tip of #2408 (ie 30c4275) - only tested the creation of a new user in a new org - no more crazy hammering of the ldap, the two ldap entries for the org are correctly created under ou=pendingorgs, the ldap entry for the user is created under ou=pendingusers as expected. Yay!

Still got the NPE from the `mailservice.EmailFactory.sendNewAccountRequiresModerationEmail` as reported in https://github.com/georchestra/georchestra/issues/2108#issuecomment-456110851 - will dig deeper to see if it's a local issue or not - is this part of the code covered by the integration tests ? `moderatedSignup=true` is set in `console.properties` so that the pending behaviour (& SUPERUSER notification) triggers.

**Comment by landryb** [comment]
`console/account/passwordRecovery` also triggers the same NPE here, so that's probably something wrong on my side with mail templating. Maybe something related to the config.jar removal and property priorities recent work..... will try a full clean rebuild.

**Comment by landryb** [comment]
Hm. So after updating datadir & cleaning everything, the `passwordRecovery` handler manages to send an email. If i then retry creating a new org/new user, a different NPE is triggered:
```
Caused by: javax.mail.SendFailedException: No recipient addresses
        at javax.mail.Transport.send0(Transport.java:148)
        at javax.mail.Transport.send(Transport.java:118)
        at org.georchestra.console.mailservice.Email.send(Email.java:175)
        at org.georchestra.console.mailservice.Email.send(Email.java:135)
        at org.georchestra.console.mailservice.EmailFactory.sendNewAccountRequiresModerationEmail(EmailFactory.java:138)
```

which is a bit puzzling, as SUPERUSER role has a member, and this member has a valid e-mail address:
```
ou=roles,~ > cat cn=SUPERUSER member

dn: cn=SUPERUSER,ou=roles,dc=georchestra,dc=org
member: uid=testadmin,ou=users,dc=georchestra,dc=org

ou=users,~ > cat uid=testadmin mail

dn: uid=testadmin,ou=users,dc=georchestra,dc=org
mail: root@craig.fr
```

I see that `memberOf` is used in `findByRole()`, which is called by the code in https://github.com/georchestra/georchestra/blob/master/console/src/main/java/org/georchestra/console/ws/newaccount/NewAccountFormController.java#L271 

Looking a bit more at my DIT, the `memberOf` overlay seems locally broken, as querying `memberOf` attr on this user doesnt show SUPERUSER:
```
ldapsearch -x -h localhost -b dc=georchestra,dc=org uid=testadmin memberOf
# testadmin, users, georchestra.org
dn: uid=testadmin,ou=users,dc=georchestra,dc=org
memberOf: cn=GN_REVIEWER,ou=roles,dc=georchestra,dc=org
memberOf: cn=GN_EDITOR,ou=roles,dc=georchestra,dc=org
memberOf: cn=shortorg,ou=orgs,dc=georchestra,dc=org
```

Probably related to the many testing i've done in this DIT. after fixing that (and properly having a user reachable via `memberOf` in SUPERUSER) by removing and readding `testadmin` from that role, creating a new account works fine, the web page doesn't trigger the NPE anymore, the requesting user gets an ACK mail, and testadmin recieves the pending acct notification; Yay !

All this makes me ponder, should there be a test somewhere in this calling code for the case of having no user with the SUPERUSER role ? Granted, this shouldnt happen, but a NPE is never nice...

**Comment by landryb** [comment]
on the 'all users' page (`console/manager/#!/browse/all/users`), pending users are listed and nothing differenciates them from regular users - i might be mistaken, but in the past (not 100% sure) pending users on this page were listed in italic ? Do we want to list pending users on the 'all users' page ? I think it's fine only having them on `/console/manager/#!/browse/pending/users`..

Same thing for orgs, pending orgs are listed on the all/orgs page and nothing makes them 'stand out' - all long name labels are in italic there.. maybe it'd be better to not show the pending orgs there ?

I note that on `/console/account/new` in the dropdown org list, only the 'validated orgs' are listed, and that's the expected behaviour imo. It would be awkward to register oneself against an org which hasnt been validated (with possibly ugly/broken labels, etc) .... unless several users of the same new org try to register in a row, and each of them tries to create the same new org at the same time ?

**Comment by landryb** [comment]
While testing, i noticed that a pending user waiting for validation is added as a member of the USER role (as `uid=username,ou=pendingusers`) and removed from the same role if rejected/deleted instead of being validated - i'm not 100% sure this is the expected behaviour.

To me only *valid* users should belong to the USER role... as `description` attribute says `This group grants basic, authenticated access to the whole SDI` so the *added to USER role* event should happen when confirming the user account, which should be added as `uid=username,ou=users`

**Comment by fvanderbiest** [comment]
Thanks for your comments & extensive testing Landry.
I'd like to close this issue (job is done on master) and open issues if required.
I'm gonna do some testing too on my side.

**Comment by landryb** [comment]
Fine with me, thanks @cmangeat for the many hours spent on this :)
